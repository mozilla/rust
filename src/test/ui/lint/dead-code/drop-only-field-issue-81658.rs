//! The field `guard` is never used directly, but it is still useful for its side effect when
//! dropped. Since rustc doesn't consider a `Drop` impl as a use, we want to make sure we at least
//! produce a helpful diagnostic that points the user to what they can do if they indeed intended to
//! have a field that is only used for its `Drop` side effect.
//!
//! Issue: https://github.com/rust-lang/rust/issues/81658

#![deny(dead_code)]

use std::sync::{Mutex, MutexGuard};

/// Holds a locked value until it is dropped
pub struct Locked<'a, T> {
    // Field is kept for its affect when dropped, but otherwise unused
    guard: MutexGuard<'a, T>, //~ ERROR field is never read
}

impl<'a, T> Locked<'a, T> {
    pub fn new(value: &'a Mutex<T>) -> Self {
        Self {
            guard: value.lock().unwrap(),
        }
    }
}

fn main() {
    let items = Mutex::new(vec![1, 2, 3]);

    // Hold a lock on items while doing something else
    let result = {
        // The lock will be released at the end of this scope
        let _lock = Locked::new(&items);

        do_something_else()
    };

    println!("{}", result);
}

fn do_something_else() -> i32 {
    1 + 1
}
