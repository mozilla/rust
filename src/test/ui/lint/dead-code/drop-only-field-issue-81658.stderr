error: field is never read: `guard`
  --> $DIR/drop-only-field-issue-81658.rs:15:5
   |
LL |     guard: MutexGuard<'a, T>,
   |     -----^^^^^^^^^^^^^^^^^^^
   |     |
   |     help: if this is intentional, prefix it with an underscore: `_guard`
   |
   = note: the leading underscore signals that this field serves some other purpose even if it isn't used in a way that we can detect. (e.g. for its effect when dropped or in foreign code)
note: the lint level is defined here
  --> $DIR/drop-only-field-issue-81658.rs:8:9
   |
LL | #![deny(dead_code)]
   |         ^^^^^^^^^

error: aborting due to previous error

